PEP 778: Add initial text and reserve 777 (#3786)
python/peps: peps/pep-0778.rst

PEP: 778
Title: Supporting Symlinks in Wheels
Author: Ethan Smith <ethan@ethanhs.me>

> Sponsor: Barry Warsaw <barry@python.org>

Reviewer: ```suggestion
Sponsor: Barry Warsaw <barry@python.org>
PEP-Delegate: Paul Moore <p.f.moore@gmail.com>
```

Status: Draft
Type: Standards Track
Topic: Packaging

> Content-Type: text/x-rst

Reviewer: This is no longer needed and no longer in the template: https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1
```suggestion
```

Author: I must have had an old checkout, thanks for correcting!

    my_package/libfoo.so.3,my_package/libfoo.so

The format of ``LINKS``, as can seen above, is ``source_path,target_path`` where ``source_path``
is a path relative to the root of any namespace or package root in the wheel. ``target_path`` is a
*non-dangling* path (i.e. a path that exists on the filesystem after the wheel's contents are
extracted) in a package or namespace of any package in the wheel. This means that if a wheel
contains multiple packages, all paths in packages in the wheel are acceptable.

Installer Behavior Specification
--------------------------------

Installers **MUST** resolve the paths of any link contained in the ``LINKS`` file *before*
deciding if any ``source_path`` or ``target_path`` are valid. Installers **MUST** verify that
``source_path`` and ``target_path`` are located inside any namespace or package coming from the
wheel.  Installers **MUST** reject cyclic symlinks in wheels. Installers **MAY** error if a long
chain of symlinks (symlinks pointing to symlinks many times repeated) exceeds a limit set by the
installer.

Installers **MUST** follow the following steps when handling a wheel with symlinks:

> 1. Check for the existance of a ``LINKS`` file in the ``.dist-info``. If it does not exist,

Reviewer: ```suggestion
1. Check for the existence of a ``LINKS`` file in the ``.dist-info``. If it does not exist,
```

logins.

This PEP lists several requirements on checks to run by installers on symlinks in wheels to ensure
attacks like the one described above cannot happen. This means it is **critical** that installers
carefully implement these security safeguards and prevent malicious use on package installation.

In particular, the following checks **MUST** be made by installers:

1. That the symlinks do not point outside of any packages or namespaces coming from the wheel
2. That the symlinks are not dangling (the target exists at install time)
3. That the symlinks are not cyclical, stopping after a certain depth of checking to avoid denial
   of service requests

Do not follow symlinks on removal.

How to Teach This
=================

End users should, once the changes have propagated through the ecosystem, transparently experience
the benefits of symlinks in wheels. It is important for installers to give clear error messages if
symlinks are unsupported on the platform, and explain why installation has failed.

For people building libraries, documentation on ``packaging.python.org`` should describe the use

> cases and caveats (especially platform suppport) of symlinks in wheels. Otherwise it should be

Reviewer: ```suggestion
cases and caveats (especially platform suppport) of symlinks in wheels. Otherwise it should be
```

Author: Was there a change you wanted to suggest here? I don't see any diff.

Reviewer: I'm not sure what happened there! Edited to remove the third P in "suppport".

files. This PEP allows for junctions as users may wish to only link folders to a different
location, and future :pep:`660` implementations may need to rely on this feature.

Put symlinks in the ``RECORD`` Metadata File
--------------------------------------------

While this could be done, it would clutter the ``RECORD`` file. Furthermore the most
straightforward implementation would place the target at the end of the record. This would
make it harder to scan across the line and visually see what symlinks exist in the wheel.

Library Maintainers Should Use Python to Locate Libraries
---------------------------------------------------------

Using Python to locate libraries would be much easier. However, some libraries like ``libtorch``
are used by extension modules and themselves require loading dependencies. Some compiled libraries
cannot use Python to find their loader dependencies.

Include Support for Hardlinks
-----------------------------

This PEP does not specify any behavior around hardlinks. This is intentional. This is left as an
extension to a future PEP.

Open Issues
===========

> PEP 660 and Deferring Editable Installation Support
> ------------------------------------------------------

Reviewer: ```suggestion
PEP 660 and Deferring Editable Installation Support
---------------------------------------------------
```

leaves that as an open question to be explored in a future PEP.

Rationale
=========

To support the 3 main namings of a library used in loading and library linking on Unix, we
propose adding support for symlinks in Python wheels. To allow for tracking symlinks made, and to
potentially support other platforms that may not support Unix symlinks directly, we propose the
use of a new wheel metadata file ``LINKS``, which will exist in the ``.dist-info`` directory alongside
``METADATA``, ``RECORD``, and other metadata files.

Using a ``LINKS`` file will allow for more cross-platform uses of symlink-like usage. On Windows,
symlinks require either `a group policy allowing the user to make symlinks
<https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-10/security/threat-protection/security-policy-settings/create-symbolic-links>`_
(e.g. by enabling `Dev Mode
<https://learn.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development>`_)
or Administrative permissions. This means that it may be the case that symlinks are unsupported on
some user systems. By using a ``LINKS`` file, installers will be able to potentially use other

> methods of handlings symlinks, such as junctions on Windows, where otherwise the installer would

Reviewer 2: ```suggestion
methods for handling symlinks, such as junctions on Windows, where otherwise the installer would
```

4. Next, check that the installer can make some kind of link for each pair in the site directory.
   If the installer cannot make a link for the file/folder ``target_path`` for the current
   platform, an error **MUST** be raised. An example of a failure mode would be a UNIX symlink to
   a file target, where the installer is running on Windows and the installer cannot make
   symlinks but can make junctions. In this case the installer **MUST** error because it cannot
   handle the link.
5. Finally, the installer **MUST** add a platform-relevant link between ``source_path`` and
   ``target_path``.

Installers **MUST NOT** by default copy files instead of generating a symlink when handling
symlinks. Installers **MAY** have such behavior available under an alternate configuration or
command line flag.

Build Backend Specification
---------------------------

When creating a wheel, build backends **MUST** treat symlinks in the same way as its target when
deciding whether to include the symlink in a wheel. Build backends **MUST** verify that there are
no dangling symlinks in the ``LINKS`` file. Build backends **SHOULD** recognize platform-relevant

> symlinks that would be included in builds. On Unix this is typically symlinks, on Windows this

Reviewer 2: ```suggestion
symlinks that would be included in builds. On POSIX systems this is typically symlinks, on Windows this
```
